source/io/module_javaScriptInvoke.js

```diff
@@ -358,30 +401,46 @@
                 return;
             }
 
-            var context = undefined;
+            var returnTypeName = getParameterTypeString(returnPointer, 0);
+            var returnValuePointer = undefined;
+            if (returnTypeName !== 'StaticTypeAndData') {
+                // User doesn't want return value. We're passing '*' for the return in VIA code, we get StaticTypeAndData
+                returnValuePointer = JavaScriptInvoke_GetParameterPointer(returnPointer, 0);
+            }
+            if (returnTypeName === 'Array') {
+                returnTypeName += getArrayElementTypeString(returnValuePointer);
+            }
+
+            var completionCallbackStatus = {
+                retrievalState: completionCallbackRetrievalEnum.AVAILABLE,
+                invocationState: completionCallbackInvocationEnum.PENDING
+            };
+
             var returnValue = undefined;
+            var context = generateContext(occurrencePointer, functionName, returnTypeName, returnValuePointer, errorStatusPointer, errorCodePointer, errorSourcePointer, completionCallbackStatus);
             try {
                 returnValue = functionToCall.apply(context, parameters);
             } catch (ex) {
                 newErrorStatus = true;
                 newErrorCode = ERRORS.kNIUnableToInvokeAJavaScriptFunction.CODE;
                 newErrorSource = Module.coreHelpers.formatMessageWithException(ERRORS.kNIUnableToInvokeAJavaScriptFunction.MESSAGE + '\'' + functionName + '\'.', ex);
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                completionCallbackStatus.retrievalState = completionCallbackRetrievalEnum.UNRETRIEVABLE;
+                completionCallbackStatus.invocationState = completionCallbackInvocationEnum.REJECTED;
+                Module.eggShell.setOccurrence(occurrencePointer);
                 return;
             }
-
-            try {
-                updateReturnValue(functionName, returnPointer, returnValue, errorStatusPointer, errorCodePointer, errorSourcePointer);
-            } catch (ex) {
-                newErrorStatus = true;
-                newErrorCode = ERRORS.kNIUnableToSetReturnValueInJavaScriptInvoke.CODE;
-                newErrorSource = Module.coreHelpers.formatMessageWithException(ERRORS.kNIUnableToSetReturnValueInJavaScriptInvoke.MESSAGE + '\'' + functionName + '\'.', ex);
-                Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+            // assume synchronous invocation since the completion callback was never retrieved
+            if (completionCallbackStatus.retrievalState === completionCallbackRetrievalEnum.AVAILABLE) {
+                updateReturnValue(functionName, returnTypeName, returnValuePointer, returnValue, errorStatusPointer, errorCodePointer, errorSourcePointer);
```

Comment:
Why was the try - catch around updateReturnValue removed?
@cglzaguilar are there known cases that can cause updateReturnValue to throw? Are they covered in tests?

If y'all think we should keep the exception handling then maybe we can do:

```
tryUpdateReturnValue = function () {

try {
updateReturnValue
} catch() {
update lv error
}
```

So we do not have to duplicate the exception logic in two places